c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -135,12 +136,35 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                     PointerCast::ArrayToPointer => {}
                     PointerCast::Unsize => {}
                 }
-                self.do_assign_pointer_ids(to_lty.label, from_lty.label)
-                // TODO add other dataflow constraints
+
+                // We would normally do `self.do_assign`,
+                // but we can't do a normal `self.do_equivalence_nested` (inside `self.do_assign`),
+                // so we do the `self.do_assign_pointer_ids` from `self.do_assign`,
+                // and then do the body of `self.do_equivalence_nested`,
+                // but after the extra outer layers are stripped.
+
+                self.do_assign_pointer_ids(to_lty.label, from_lty.label);
+
+                // Each of these ptr casts needs at least the outer layer stripped (e.x. `&`, `*mut`, `fn`),
+                // but then after that, these ptr casts below
+                // needs more layers stripped to reach the types that are equivalent.
+                let strip_from = |lty: LTy<'tcx>| match ptr_cast {
+                    PointerCast::ArrayToPointer => lty.args[0],
+                    PointerCast::Unsize => lty.args[0],
+                    _ => lty,
+                };
+                let strip_to = |lty: LTy<'tcx>| match ptr_cast {
+                    PointerCast::Unsize => lty.args[0],
+                    _ => lty,
+                };
+                for (&from_lty, &to_lty) in iter::zip(from_lty.args, to_lty.args) {
+                    self.do_unify(strip_from(from_lty), strip_to(to_lty));
+                }
```

Comment:
Similar to my previous comment, I think it would be good to be clear about how many args we expect to see for each `PointerCast` case.  For example, `Unsize` should always have 1 arg in both `from_lty` and `to_lty`, while for `UnsafeFnPointer` we only expect that `from_lty` and `to_lty` will have the same amount.

I also think it's okay to panic on the tricky cases `ReifyFnPointer` and `ClosureFnPointer` for now.

Reply:
How tricky is `ReifyFnPointer`?  If it's not too complex, I'd rather not error on it as it is used commonly in `lighttpd` and other `c2rust transpile`d code for `fn` ptrs.

Reply:
With `ReifyFnPointer`, I haven't thought about what constraints we ought to generate for pointer arguments in the function's signature.  For example, given `fn f(p: *mut /*p0*/ i32)` and the expression `f as fn(*mut /*p1*/ i32)`, do we require `p1`'s permissions to be a subset of `p0`'s, or do we require the reverse, or do we require the two to be equal?